### PR TITLE
[REVIEW] adding notebooks and utilities for the mortgage workflow

### PR DESCRIPTION
**Pull request by mt-jones**

### CODE REVIEW

File: utils/conda_create_cudf.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+echo -e "\n"                                                                            
+echo "attempting to remove current conda environment cudf"                              
+conda-env   remove  --name cudf  --quiet --yes                                          
+echo "creating dev environment for cudf"                                                
+echo -e "\n"                                                                            
+conda       update  --name base --yes conda                                          && \
+conda       install --yes  python=$PYTHON_VERSION                                    && \
+conda       create  --name cudf  --yes python=$PYTHON_VERSION                        && \
+conda       install --name cudf  --yes --channel conda-forge                            \
```
Comment (randerzander) [inline utils/conda_create_cudf.sh]: I might be misreading, but this looks like a conda environment for building cudf.

I'd think we want to just conda install cudf, cuml, nvstrings, dask, dask_cudf, dask_xgboost (+ other packages?) here, no?

**Comment by randerzander** [comment]
The notebook assumes the mortgage dataset is already available in `/datasets/` -- should the notebook include a cell that downloads and saves them? Or include a README note instructing how the user can manually download/prepare them?

File: utils/conda_create_cudf.sh (hunk 1 above)
Comment (mt-jones) [reply utils/conda_create_cudf.sh]: These utilities are not meant to install any of the RAPIDS projects, but to prepare a user's environment to build them from source, or install with Conda.

**Comment by mt-jones** [comment]
> The notebook assumes the mortgage dataset is already available in `/datasets/` -- should the notebook include a cell that downloads and saves them? Or include a README note instructing how the user can manually download/prepare them?

https://rapidsai.github.io/datasets/

I can add an explicit reference to this site, but the path `/datasets` was meant to be a placeholder. I assume the user will change this path.

File: utils/conda-create-cudf.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+conda-env   remove  --name cudf  --quiet --yes                                          
+echo "creating dev environment for cudf"                                                
+echo -e "\n"                                                                            
+conda       update  --name base --yes conda                                          && \
+conda       install --yes  python=$PYTHON_VERSION                                    && \
+conda       create  --name cudf  --yes python=$PYTHON_VERSION                        && \
+conda       install --name cudf  --yes --channel conda-forge                            \
+                                      --channel numba                                   \
+                                      --channel nvidia                                  \
+                                      nvstrings                                         \
```
Comment (harrism) [inline utils/conda-create-cudf.sh]: Typically the dependency list is specified (with minimum / maximum versions) in a .yml file.

File: mortgage/E2E.ipynb [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+    "        (\"num_units\", \"int64\"),\n",
+    "        (\"occupancy_status\", \"category\"),\n",
+    "        (\"property_state\", \"category\"),\n",
+    "        (\"zip\", \"int64\"),\n",
+    "        (\"mortgage_insurance_percent\", \"float64\"),\n",
+    "        (\"product_type\", \"category\"),\n",
+    "        (\"coborrow_credit_score\", \"float64\"),\n",
+    "        (\"mortgage_insurance_type\", \"float64\"),\n",
+    "        (\"relocation_mortgage_indicator\", \"category\")\n",
+    "    ])\n",
```
Comment (harrism) [inline mortgage/E2E.ipynb]: Are you missing this line? `("quarter", "int64")`

File: utils/README.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
+# Utility Scripts
+
+## Quick Summary
+
+* `conda-create-cudf.sh`: creates a conda environment named `cudf` with all of the requisite software dependencies for [cuDF](https://github.com/rapidsai/cudf)
+* `start-jupyter.sh`: starts a JupyterLab environment for interacting with, and running, notebooks
+* `stop-jupyter.sh`: identifies all process IDs associated with Jupyter and kills them
+* `dask-cluster.py`: launches a configured Dask cluster (a set of nodes) for use within a notebook
+* `dask-setup.sh`: a low-level script for constructing a set of Dask workers on a single node ... **do not use this script directly**
```
[the remark this answers is not in the document]
Comment (mt-jones) [reply utils/README.md]: It should be called either by `dask-cluster.py` or by the cell within the notebook.

I want to discourage users from editing this file. Maybe I should just write this, instead.

File: mortgage/E2E.ipynb (hunk 3 above)
Comment (mt-jones) [reply mortgage/E2E.ipynb]: This is missing. Thanks for the catch.

**Comment by randerzander** [comment]
> > The notebook assumes the mortgage dataset is already available in `/datasets/` -- should the notebook include a cell that downloads and saves them? Or include a README note instructing how the user can manually download/prepare them?
> 
> https://rapidsai.github.io/datasets/
> 
> I can add an explicit reference to this site, but the path `/datasets` was meant to be a placeholder. I assume the user will change this path.

Do @harrism 's READMEs or any cells instruct the user how to download and prepare this dataset?

**Comment by harrism** [comment]
My README says "3.  Download the mortgage dataset, following the instructions provided at https://rapidsai.github.io/demos/docs/datasets/2018-12-05-mortgage-dataset/

     You will need to update paths and years in the notebook (see below) depending on which subset of the mortgage data you download and where you install it."

When you click that link, there are instructions.

**Comment by randerzander** [comment]
> My README says "3. Download the mortgage dataset, following the instructions provided at https://rapidsai.github.io/demos/docs/datasets/2018-12-05-mortgage-dataset/
> 
> You will need to update paths and years in the notebook (see below) depending on which subset of the mortgage data you download and where you install it."
> 
> When you click that link, there are instructions.

Thanks. @mt-jones since a lot of people just try to run notebooks as is, can you put a comment in the cell that first uses /datasets, and ask the user to point it to wherever they saved the dataset? Bonus point if you link to the dataset too..

File: README.md [hunk 5]
```diff
@@ -1,2 +1,6 @@
-# notebooks
-RAPIDS Sample Notebooks
+# RAPIDS Notebooks and Utilities
+
+## Quick Summary
```
[the remark this answers is not in the document]
Comment (mt-jones) [reply README.md]: Fixed this.

File: utils/conda-create-cudf.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (mt-jones) [reply utils/conda-create-cudf.sh]: Removed.

**Pull request merged**